### PR TITLE
Add proptools.FilterPropertyStruct

### PR DESCRIPTION
**Pull request by colincross**
Move some code from Soong to support creating a property struct
at runtime by filtering fields out of another property struct.

Test: TestFilterPropertyStruct
Change-Id: Ic5ae390a885195bebad6f3ecb7c752c0582a60b1

### CODE REVIEW

File: proptools/filter.go [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+			if field.Type.Elem().Kind() == reflect.Struct {
+				nestedType, subFiltered := filterPropertyStruct(field.Type.Elem(), subPrefix, predicate)
+				filtered = filtered || subFiltered
+				if nestedType == nil {
+					continue
+				}
+				field.Type = reflect.PtrTo(nestedType)
+			}
+		case reflect.Interface:
+			panic("Interfaces are not supported in arch_variant properties")
```
Comment (gyias) [inline proptools/filter.go]: Isn't arch_variant a Soong concept?

File: proptools/filter.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+	}
+
+	return ret, true
+}
+
+// FilterPropertyStructSharded takes a reflect.Type that is either a sturct or a pointer to a struct, and returns a list
+// of reflect.Type that only contains the fields in the original type for which predicate returns true, and a bool that
+// is true if the new struct type has fewer fields than the original type.  If there are no fields in the original type
+// for which predicate returns true it returns nil and true.  Each returned struct type will have a maximum of 10 top
+// level fields in it to attempt to avoid hitting the reflect.StructOf name length limit, although the limit can still
```
Comment (gyias) [inline proptools/filter.go]: Q: What are you referring to by "reflect.StructOf name length limit"? My quick Google search failed to find any relevant info.

File: proptools/filter.go (hunk 1 above)
Comment (colincross) [reply proptools/filter.go]: Fixed

File: proptools/filter.go (hunk 2 above)
Comment (colincross) [reply proptools/filter.go]: Reworded, it hits a panic at https://golang.org/src/reflect/type.go#L537.

**Pull request merged**